{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collection of siRNA Features from the Input Sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Packages Installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gffutils in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (0.11.0)\n",
      "Requirement already satisfied: argcomplete>=1.9.4 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from gffutils) (2.0.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from gffutils) (1.16.0)\n",
      "Requirement already satisfied: simplejson in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from gffutils) (3.17.6)\n",
      "Requirement already satisfied: argh>=0.26.2 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from gffutils) (0.26.2)\n",
      "Requirement already satisfied: pyfaidx>=0.5.5.2 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from gffutils) (0.7.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from pyfaidx>=0.5.5.2->gffutils) (58.0.4)\n",
      "Requirement already satisfied: pyfaidx in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (0.7.0)\n",
      "Requirement already satisfied: setuptools>=0.7 in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from pyfaidx) (58.0.4)\n",
      "Requirement already satisfied: six in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from pyfaidx) (1.16.0)\n",
      "Requirement already satisfied: biopython in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (1.79)\n",
      "Requirement already satisfied: numpy in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (from biopython) (1.23.0)\n",
      "Requirement already satisfied: seqfold in /Users/christian/opt/anaconda3/lib/python3.9/site-packages (0.7.14)\n"
     ]
    }
   ],
   "source": [
    "%%sh\n",
    "pip install gffutils     # https://pythonhosted.org/pyfaidx/\n",
    "pip install pyfaidx      # https://pythonhosted.org/gffutils/contents.html\n",
    "pip install biopython    # http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
    "pip install seqfold      # https://pypi.org/project/seqfold/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Parsing FASTA file for being BioPython readable and for features extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Parsing process + we are already thresholding the scores specific cutoff\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "Scores = []\n",
    "for seqs in SeqIO.parse(\"data.fasta\", \"fasta\"):\n",
    "    score = ((float(seqs.id[5:])))\n",
    "    if score >= 0.666:\n",
    "        Scores.append(float(1))\n",
    "    else:\n",
    "        Scores.append(float(0))\n",
    "print(Scores)\n",
    "print(len(Scores))\n",
    "#the score represent the experimentally validated one that I put in the fasta header in order \n",
    "#to have an easy way to extract it\n",
    "#here I would need to add some line for the exclusion of those sequences that contains toxic motif as described in Lawlor et al., 2012, Krzyzosiak et al., 2012, Ahmed et al., 2011"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.462, 0.384, 0.514, 0.364, 0.522, 0.442, 0.441, 0.436, 0.591, 0.511, 0.555, 0.547, 0.44, 0.651, 0.759, 0.624, 0.858, 0.849, 0.383, 0.537, 0.832, 0.684, 0.736, 0.83, 0.389, 0.757, 0.552, 0.773, 0.707, 0.518, 0.753, 0.817, 0.794, 0.753, 0.836, 0.76, 0.423, 0.788, 0.871, 0.761, 0.821, 0.72, 0.863, 0.471, 0.868, 0.776, 0.849, 0.87, 0.544, 0.856, 0.835, 0.746, 0.825, 0.849, 0.754, 0.83, 0.928, 0.616, 0.825, 0.808, 0.804, 0.723, 0.597, 0.757, 0.887, 0.688, 0.79, 0.627, 0.69, 0.827, 0.893, 0.372, 0.517, 0.336, 0.268, 0.255, 0.737, 0.59, 0.652, 0.524, 0.534, 0.43, 0.584, 0.349, 0.766, 0.743, 0.565, 0.412, 0.829, 0.364, 0.794, 0.552, 0.459, 0.682, 0.7, 0.232, 0.584, 0.423, 0.813, 0.449, 0.721, 0.722, 0.578, 0.726, 0.393, 0.254, 0.442, 0.251, 0.367, 0.659, 0.449, 0.58, 0.511, 0.298, 0.598, 0.737, 0.417, 0.74, 0.727, 0.643, 0.755, 0.673, 0.443, 0.476, 0.683, 0.924, 0.82, 0.624, 0.862, 0.445, 0.565, 0.958, 0.422, 0.952, 0.945, 0.984, 0.961, 0.236, 0.27, 0.642, 0.306, 0.806, 0.92, 0.897, 0.845, 0.903, 1.002, 0.556, 0.433, 0.981, 0.792, 0.952, 0.996, 0.974, 0.576, 0.32, 0.762, 0.931, 0.488, 0.744, 0.7, 0.833, 0.684, 0.866, 0.83, 0.809, 0.644, 0.945, 0.631, 0.876, 0.949, 0.999, 0.85, 0.623, 0.762, 0.923, 0.554, 1.045, 0.995, 0.8, 0.977, 0.486, 0.444, 0.641, 0.52, 0.352, 0.758, 0.602, 1.003, 0.853, 0.949, 0.907, 0.447, 0.661, 0.95, 0.883, 0.937, 0.329, 0.5, 1.003, 1.015, 0.878, 0.765, 0.616, 0.492, 0.778, 0.827, 0.463, 0.787, 0.836, 0.735, 0.842, 0.84, 0.804, 0.838, 0.775, 0.751, 0.714, 0.712, 0.51, 0.759, 0.687, 0.499, 0.727, 0.743, 0.815, 0.531, 0.624, 0.608, 0.505, 0.847, 0.839, 0.836, 0.744, 0.841, 0.648, 0.863, 0.722, 0.5, 0.772, 0.479, 0.467, 0.635, 0.694, 0.757, 0.626, 0.73, 0.46, 0.635, 0.846, 0.482, 0.697, 0.618, 0.777, 0.801, 0.825, 0.853, 0.637, 0.723, 0.475, 0.467, 0.862, 0.781, 0.798, 0.79, 0.817, 0.532, 0.846, 0.835, 0.55, 0.564, 0.483, 0.508, 0.742, 0.728, 0.618, 0.767, 0.611, 0.455, 0.782, 0.851, 0.581, 0.543, 0.677, 0.612, 0.766, 0.612, 0.676, 0.641, 0.699, 0.671, 0.664, 0.654, 0.659, 0.686, 0.631, 0.628, 0.645, 0.613, 0.561, 0.697, 0.771, 0.802, 0.67, 0.497, 0.822, 0.767, 0.606, 0.685, 0.541, 0.804, 0.795, 0.625, 0.654, 0.797, 0.713, 0.861, 0.618, 0.921, 0.817, 0.738, 0.793, 0.564, 0.699, 0.539, 0.511, 0.851, 0.873, 0.689, 0.759, 0.827, 0.779, 0.803, 0.575, 0.589, 0.767, 0.77, 0.656, 0.793, 0.605, 0.703, 0.718, 0.719, 0.66, 0.564, 0.717, 0.699, 0.517, 0.53, 0.61, 0.589, 0.649, 0.736, 0.53, 0.599, 0.417, 0.594, 0.857, 0.591, 0.538, 0.43, 0.363, 0.433, 0.295, 0.618, 0.82, 0.754, 0.817, 0.55, 0.683, 0.667, 0.797, 0.579, 0.532, 0.65, 0.652, 0.75, 0.824, 0.759, 0.609, 0.724, 0.568, 0.729, 0.49, 0.719, 0.711, 0.627, 0.836, 0.991, 0.651, 0.561, 0.792, 0.595, 0.774, 0.702, 0.68, 0.558, 0.653, 0.627, 1.016, 0.845, 0.723, 0.742, 0.772, 0.422, 0.801, 0.574, 0.494, 0.564, 0.779, 0.916, 0.758, 0.707, 0.617, 0.787, 0.793, 0.768, 0.654, 0.711, 0.758, 0.819, 0.758, 0.699, 0.868, 0.893, 0.247, 0.895, 0.986, 0.695, 0.886, 0.99, 0.564, 0.732, 0.617, 0.127, 0.397, 0.53, 0.52, 0.462, 0.402, 0.505, 0.738, 0.513, 0.67, 0.397, 0.494, 0.463, 0.576, 0.564, 0.73, 0.589, 0.602, 0.723, 0.595, 0.357, 0.218, 0.468, 0.731, 0.721, 0.749, 0.272, 0.367, 0.727, 0.615, 0.659, 0.68, 0.483, 0.585, 0.687, 0.677, 0.743, 0.584, 0.492, 0.604, 0.661, 0.293, 0.292, 0.259, 0.668, 0.521, 0.725, 0.395, 0.693, 0.72, 0.584, 0.396, 0.702, 0.278, 0.873, 0.739, 0.639, 0.774, 0.758, 0.344, 0.342, 0.438, 0.605, 0.535, 0.651, 0.709, 0.844, 0.221, 0.637, 0.663, 0.607, 0.787, 0.426, 0.719, 0.882, 0.419, 0.705, 0.547, 0.768, 0.474, 0.444, 0.606, 0.419, 0.611, 0.777, 0.409, 0.807, 0.737, 0.705, 0.459, 0.547, 0.49, 0.978, 0.753, 0.719, 0.532, 0.679, 0.492, 0.518, 0.447, 0.371, 0.584, 0.564, 0.447, 0.374, 0.669, 0.506, 0.717, 0.702, 0.523, 0.288, 0.632, 0.39, 0.621, 0.608, 0.683, 0.722, 0.725, 0.576, 0.751, 0.579, 0.908, 0.837, 0.464, 0.73, 0.625, 0.402, 0.497, 0.709, 0.605, 0.618, 0.63, 0.744, 0.666, 0.544, 0.695, 0.578, 0.527, 0.631, 0.467, 0.644, 0.291, 0.67, 0.625, 0.441, 0.658, 0.471, 0.485, 0.693, 0.345, 0.514, 0.479, 0.557, 0.646, 0.46, 0.449, 0.385, 0.424, 0.747, 0.752, 0.662, 0.649, 0.608, 0.598, 0.654, 0.508, 0.436, 0.453, 0.536, 0.5, 0.549, 0.679, 0.487, 0.677, 0.513, 0.537, 0.667, 0.249, 0.742, 0.499, 0.805, 0.829, 0.696, 0.683, 0.563, 0.686, 0.734, 0.559, 0.574, 0.758, 0.547, 0.541, 0.604, 0.639, 1.016, 0.701, 0.613, 0.49, 0.623, 0.6, 0.376, 0.517, 0.463, 0.371, 0.809, 0.775, 0.771, 0.614, 0.873, 0.666, 0.562, 0.846, 0.87, 0.6, 0.417, 0.73, 0.718, 0.828, 0.517, 0.525, 0.91, 0.743, 0.527, 0.32, 0.645, 0.594, 0.57, 0.731, 0.828, 0.501, 0.655, 0.834, 0.659, 0.718, 0.9, 1.099, 0.812, 0.697, 0.794, 0.619, 0.587, 0.753, 0.462, 0.967, 1.115, 0.614, 0.719, 0.772, 0.65, 0.747, 0.985, 0.919, 0.668, 0.719, 0.776, 0.705, 0.786, 0.931, 0.951, 0.777, 1.039, 0.864, 0.825, 0.957, 0.614, 0.553, 0.91, 0.866, 0.494, 0.669, 0.502, 0.362, 0.666, 0.482, 0.416, 0.842, 0.486, 0.783, 0.684, 1.1, 0.958, 0.81, 0.86, 0.465, 1.058, 0.984, 1.001, 0.999, 0.64, 1.064, 0.679, 0.953, 0.964, 0.992, 1.016, 0.779, 1.084, 0.642, 0.5, 0.836, 1.016, 0.678, 0.995, 0.34, 0.917, 0.638, 0.955, 0.756, 0.967, 0.841, 0.573, 0.777, 0.98, 0.711, 0.702, 0.715, 0.929, 0.678, 0.838, 0.895, 0.758, 0.839, 0.489, 0.888, 0.521, 0.719, 0.904, 0.744, 0.874, 0.601, 0.56, 0.859, 0.705, 0.76, 0.801, 0.592, 0.478, 0.511, 0.596, 0.451, 0.496, 0.789, 0.883, 0.591, 0.81, 0.892, 0.75, 0.539, 0.864, 0.59, 0.527, 0.652, 0.873, 0.502, 0.681, 0.922, 0.732, 0.825, 0.78, 0.524, 0.721, 0.893, 0.631, 0.456, 0.531, 0.549, 0.839, 0.726, 0.722, 0.588, 0.345, 0.782, 0.844, 0.818, 0.55, 0.488, 0.516, 0.569, 0.854, 0.798, 0.824, 0.713, 0.572, 0.457, 0.428, 0.45, 0.44, 0.494, 0.585, 0.579, 0.43, 0.415, 0.43, 0.364, 0.491, 0.42, 0.548, 0.701, 0.55, 0.47, 0.498, 0.534, 0.567, 0.4, 0.563, 0.778, 0.707, 0.76, 0.694, 0.443, 0.485, 0.715, 0.706, 0.395, 0.394, 0.557, 0.635, 0.593, 0.855, 0.523, 0.631, 0.745, 0.678, 0.856, 0.544, 0.449, 0.316, 0.604, 0.917, 0.815, 0.547, 0.489, 0.463, 0.552, 0.404, 0.539, 0.54, 0.423, 0.585, 0.499, 0.365, 0.252, 0.465, 0.446, 0.359, 0.451, 0.54, 0.357, 0.416, 0.339, 0.53, 0.573, 0.565, 0.664, 0.472, 0.617, 0.515, 0.436, 0.887, 0.611, 0.553, 0.626, 0.599, 0.516, 0.577, 1.077, 0.978, 0.854, 0.548, 0.796, 0.586, 0.438, 0.805, 0.716, 0.818, 0.597, 0.559, 0.601, 0.5, 0.899, 0.59, 0.691, 0.482, 0.742, 0.318, 0.896, 0.777, 0.778, 0.749, 0.829, 0.657, 0.613, 0.74, 0.713, 0.676, 0.649, 0.374, 0.53, 0.908, 0.436, 0.458, 0.51, 0.859, 0.952, 0.867, 0.712, 0.537, 0.846, 0.43, 0.381, 0.523, 0.709, 0.516, 0.698, 0.64, 0.556, 0.635, 0.495, 0.799, 0.737, 0.625, 0.969, 0.895, 0.742, 0.971, 1.072, 0.873, 0.776, 0.761, 0.738, 0.883, 0.691, 0.93, 0.863, 1.071, 0.732, 0.81, 0.714, 0.485, 0.352, 0.605, 0.699, 0.538, 0.538, 0.617, 0.679, 0.772, 0.569, 0.988, 1.06, 1.227, 0.657, 1.173, 0.999, 0.623, 1.081, 0.501, 1.126, 0.699, 1.026, 1.201, 0.801, 0.616, 0.617, 0.449, 0.522, 0.782, 0.692, 1.146, 0.957, 1.216, 0.669, 0.753, 0.435, 0.96, 0.314, 0.568, 0.977, 0.322, 0.618, 0.835, 0.663, 0.719, 0.761, 0.78, 0.433, 0.864, 0.689, 0.723, 0.784, 0.746, 0.788, 0.889, 0.991, 0.752, 1.066, 0.931, 1.005, 0.859, 0.804, 0.566, 1.016, 1.06, 0.919, 0.655, 0.83, 1.08, 0.803, 0.635, 0.97, 1.01, 0.994, 0.859, 1.13, 0.916, 0.951, 0.738, 0.971, 0.775, 0.95, 0.76, 0.859, 1.14, 0.756, 0.835, 0.912, 1.007, 0.527, 1.045, 0.794, 0.492, 1.145, 0.848, 1.026, 1.062, 0.804, 0.656, 0.605, 0.849, 0.438, 0.685, 0.533, 0.719, 0.808, 0.75, 0.799, 0.961, 0.878, 0.954, 0.674, 1.013, 1.071, 0.568, 1.068, 1.093, 1.004, 1.051, 0.565, 0.631, 0.874, 0.785, 1.141, 0.754, 0.794, 0.899, 0.925, 0.747, 0.574, 0.871, 0.9, 0.716, 1.021, 1.008, 0.953, 0.632, 0.758, 0.873, 0.908, 1.013, 1.057, 1.146, 1.039, 1.252, 0.887, 0.967, 0.964, 0.827, 1.098, 1.121, 1.03, 1.103, 0.675, 0.932, 0.848, 1.07, 1.042, 0.85, 1.215, 1.088, 0.931, 0.805, 0.884, 0.647, 0.733, 0.641, 0.839, 0.662, 0.966, 0.859, 0.962, 1.043, 0.788, 1.054, 1.126, 1.083, 0.73, 1.04, 0.535, 0.82, 1.151, 1.078, 0.944, 0.912, 1.032, 0.95, 0.784, 0.857, 0.747, 0.814, 1.235, 0.799, 0.88, 0.835, 0.961, 0.975, 1.111, 0.78, 0.932, 0.861, 0.949, 0.389, 1.104, 1.152, 0.993, 0.926, 1.118, 0.768, 0.736, 1.091, 1.135, 1.156, 0.75, 0.689, 0.964, 1.003, 0.993, 1.1, 0.916, 0.959, 0.988, 0.764, 0.624, 0.834, 0.558, 0.76, 0.731, 0.788, 0.474, 0.353, 0.785, 0.479, 0.562, 0.751, 0.476, 0.675, 0.887, 0.753, 0.796, 0.605, 0.544, 0.56, 0.567, 0.731, 0.504, 0.705, 0.552, 0.707, 0.61, 0.417, 0.287, 0.547, 0.748, 0.914, 0.647, 0.641, 0.726, 0.56, 0.563, 0.463, 0.84, 0.681, 0.536, 0.518, 0.48, 0.552, 0.646, 0.766, 0.927, 0.738, 0.661, 0.739, 0.701, 0.778, 0.879, 0.921, 0.874, 0.655, 0.921, 0.93, 0.899, 0.676, 0.68, 0.667, 0.749, 0.825, 0.818, 0.85, 0.62, 0.854, 0.892, 0.633, 0.857, 0.871, 0.474, 0.576, 0.442, 0.484, 0.573, 0.792, 0.871, 0.587, 0.789, 0.812, 0.71, 0.555, 0.749, 0.532, 0.616, 0.673, 0.76, 0.812, 0.897, 0.73, 0.723, 0.724, 0.763, 0.785, 0.793, 0.754, 0.681, 0.827, 0.695, 0.642, 0.81, 0.728, 0.739, 0.669, 0.746, 0.698, 0.561, 0.654, 0.694, 0.548, 0.742, 0.476, 0.895, 0.551, 0.913, 0.844, 0.813, 0.857, 0.882, 0.889, 0.458, 0.531, 0.517, 0.52, 0.522, 0.468, 0.427, 0.674, 0.608, 0.512, 0.447, 0.464, 0.646, 0.37, 0.539, 0.671, 0.886, 0.766, 0.475, 0.385, 0.57, 0.431, 0.397, 0.726, 0.515, 0.574, 0.76, 0.647, 0.436, 0.506, 0.768, 0.567, 0.57, 0.619, 0.809, 0.386, 0.703, 0.337, 0.59, 0.617, 0.323, 0.614, 0.649, 0.813, 0.723, 0.728, 0.628, 0.537, 0.236, 0.403, 0.335, 0.484, 0.687, 0.55, 0.66, 0.644, 0.552, 0.353, 0.794, 0.423, 0.517, 0.504, 0.768, 0.785, 0.784, 0.349, 0.695, 0.646, 0.662, 0.715, 0.542, 0.495, 0.605, 0.805, 0.742, 0.852, 0.613, 0.693, 0.738, 1.003, 0.918, 0.953, 0.904, 0.602, 0.917, 0.611, 0.903, 0.841, 0.775, 0.94, 0.854, 0.78, 0.439, 0.402, 0.432, 0.642, 0.498, 0.342, 0.302, 0.779, 0.613, 0.702, 0.859, 0.784, 0.587, 0.635, 0.542, 0.863, 0.622, 0.594, 0.482, 0.345, 0.355, 0.571, 0.806, 0.856, 0.759, 0.528, 0.94, 0.938, 0.694, 0.464, 0.492, 0.816, 0.921, 0.795, 0.64, 0.447, 0.369, 0.83, 0.875, 0.848, 0.961, 0.436, 0.299, 0.872, 0.706, 0.502, 0.541, 0.471, 0.506, 0.395, 0.398, 0.471, 0.624, 0.871, 0.658, 0.826, 0.883, 0.847, 0.816, 0.782, 0.614, 0.974, 0.36, 0.607, 0.943, 0.79, 0.939, 0.435, 0.564, 0.742, 1.02, 0.859, 0.727, 0.638, 0.344, 1.042, 0.935, 0.975, 0.939, 0.811, 0.719, 0.898, 0.61, 0.476, 0.54, 0.785, 0.654, 0.669, 0.544, 0.893, 0.861, 0.853, 0.412, 0.621, 0.629, 0.64, 0.425, 0.378, 0.359, 0.326, 0.359, 0.493, 0.392, 0.318, 0.693, 0.877, 0.55, 0.786, 0.76, 0.764, 0.295, 0.748, 0.947, 0.85, 0.731, 0.734, 0.734, 0.501, 0.616, 0.696, 0.879, 0.395, 0.624, 0.847, 0.842, 0.66, 0.787, 0.752, 0.361, 0.447, 0.876, 0.71, 0.931, 0.683, 0.677, 0.867, 1.032, 0.405, 0.975, 0.606, 0.517, 0.833, 0.654, 0.552, 0.78, 0.769, 0.888, 0.871, 0.909, 0.887, 0.598, 0.758, 0.563, 0.711, 0.503, 0.478, 0.452, 0.86, 0.734, 0.948, 0.924, 0.343, 0.567, 0.692, 0.72, 0.438, 0.589, 0.438, 0.756, 0.49, 0.834, 0.585, 0.747, 0.705, 0.699, 0.661, 0.697, 0.734, 0.425, 0.492, 0.914, 0.507, 0.588, 0.642, 0.287, 0.382, 0.576, 0.688, 0.501, 0.368, 0.684, 0.595, 0.528, 0.905, 0.758, 0.52, 0.672, 0.732, 0.473, 0.511, 0.343, 0.626, 0.816, 0.82, 0.805, 0.885, 0.874, 0.423, 0.712, 0.815, 0.828, 0.359, 0.524, 0.698, 0.842, 0.901, 0.321, 0.893, 0.89, 0.692, 0.547, 0.53, 0.552, 0.466, 0.753, 0.951, 0.945, 0.788, 0.841, 0.826, 0.519, 0.715, 0.898, 0.47, 0.777, 0.49, 0.7, 0.9, 0.82, 0.897, 0.63, 0.714, 0.875, 0.921, 0.909, 0.903, 0.844, 0.841, 0.835, 0.568, 0.638, 0.388, 0.829, 0.894, 0.961, 0.631, 0.765, 0.571, 0.705, 0.9, 0.527, 0.95, 0.888, 0.58, 0.92, 0.963, 0.388, 0.73, 0.828, 0.609, 0.888, 0.923, 0.586, 0.861, 0.538, 0.599, 0.523, 0.953, 0.856, 0.845, 0.96, 0.91, 0.887, 0.656, 0.814, 0.95, 0.458, 0.837, 0.96, 0.365, 0.494, 0.43, 0.535, 0.714, 0.78, 0.623, 0.858, 0.954, 0.904, 0.728, 0.953, 0.928, 0.943, 0.859, 0.915, 0.936, 0.706, 0.725, 0.855, 0.435, 0.619, 0.662, 0.725, 0.422, 0.556, 0.86, 0.817, 0.402, 0.467, 0.623, 0.458, 0.765, 0.715, 0.456, 0.677, 0.444, 0.507, 0.812, 0.329, 0.679, 0.857, 0.737, 0.358, 0.93, 0.907, 0.708, 0.598, 0.845, 0.865, 0.802, 0.488, 0.341, 0.373, 0.486, 0.84, 0.79, 0.9, 0.855, 0.943, 0.552, 0.952, 0.854, 0.949, 0.634, 0.767, 0.908, 0.814, 0.875, 0.858, 0.584, 0.579, 0.968, 0.89, 1.042, 0.769, 0.667, 0.542, 0.393, 0.913, 0.703, 0.845, 0.857, 0.468, 0.663, 0.791, 0.741, 0.663, 0.486, 0.763, 0.848, 0.555, 0.543, 0.727, 0.848, 0.498, 0.628, 0.52, 0.422, 0.468, 0.992, 0.98, 1.047, 0.547, 0.776, 0.332, 0.6, 0.448, 0.552, 0.887, 0.659, 0.43, 0.508, 0.268, 0.231, 0.327, 0.668, 0.387, 0.476, 0.503, 0.702, 0.699, 0.7, 0.627, 0.353, 0.896, 1.02, 0.667, 0.654, 0.757, 0.701, 0.722, 0.354, 0.337, 0.322, 0.331, 0.724, 0.819, 0.432, 0.488, 0.693, 0.546, 0.797, 0.696, 0.449, 0.197, 0.14, 0.708, 0.662, 0.514, 0.589, 0.356, 0.748, 0.626, 0.765, 0.826, 0.883, 1.013, 0.854, 0.825, 0.862, 0.826, 0.625, 0.856, 0.88, 0.904, 0.933, 0.704, 0.631, 0.618, 0.824, 0.729, 0.791, 0.686, 0.455, 0.5, 0.815, 0.776, 0.533, 0.685, 0.871, 0.456, 0.958, 0.669, 0.526, 0.867, 0.681, 0.58, 0.993, 0.722, 0.689, 0.875, 1.064, 0.86, 0.858, 0.872, 0.797, 0.884, 0.787, 0.631, 0.239, 0.656, 0.749, 0.609, 0.732, 0.72, 0.915, 0.829, 0.81, 0.553, 0.779, 0.395, 0.353, 0.599, 0.691, 0.764, 0.655, 0.428, 0.676, 0.648, 0.615, 0.948, 0.843, 0.395, 0.753, 0.542, 0.796, 0.524, 0.467, 0.488, 0.549, 0.44, 0.576, 0.707, 0.316, 0.192, 0.49, 0.815, 0.609, 0.5, 0.802, 0.736, 0.561, 0.849, 0.591, 0.348, 0.152, 0.634, 0.712, 0.399, 0.821, 0.462, 0.541, 0.395, 0.164, 0.737, 0.407, 0.94, 0.632, 0.834, 0.909, 0.637, 0.768, 0.77, 0.442, 0.794, 0.577, 0.758, 0.715, 0.987, 1.119, 0.929, 0.925, 0.547, 1.153, 1.144, 1.025, 1.074, 0.552, 1.048, 1.164, 0.996, 0.814, 0.738, 0.876, 0.731, 0.618, 0.633, 0.414, 0.622, 1.048, 0.376, 0.536, 0.337, 0.911, 0.731, 0.704, 0.894, 0.946, 0.698, 0.657, 0.618, 0.811, 0.525, 0.465, 0.326, 0.905, 0.631, 0.419, 0.874, 0.758, 1.017, 1.085, 1.184, 0.721, 0.902, 0.685, 0.361, 0.508, 0.11, 0.326, 0.591, 1.008, 0.537, 0.897, 0.338, 0.545, 0.835, 0.105, 0.85, 0.002, 0.942, 1.018, 0.905, 0.973, 0.268, 1.053, 0.631, 0.755, 0.994, 1.027, 0.959, 0.408, 0.967, 1.095, 0.657, 0.812, 1.05, 1.069, 0.943, 0.729, 1.221, 0.802, 1.194, 1.001, 1.15, 1.04, 0.913, 0.939, 0.856, 0.607, 0.475, 0.94, 0.613, 0.861, 0.975, 1.15, 1.122, 0.887, 0.019, 1.028, 0.538, 0.018, 0.975, 1.037, 0.834, 1.161, 0.83, 0.323, 1.101, 0.223, 0.382, 0.871, 0.457, 0.957, 1.133, 1.08, 1.262, 1.1, 1.308, 1.1, 0.761, 0.744, 1.269, 0.862, 0.796, 0.284, 1.162, 1.316, 1.341, 0.66, 0.756, 1.041, 0.936, 0.468, 0.911, 0.916, 0.797, 0.405, 0.329, 0.944, 0.83, 0.722, 0.642, 0.66, 0.679, 0.936, 0.643, 0.507, 0.538, 0.587, 0.613, 0.626, 0.647, 0.75, 0.618, 0.808, 0.662, 0.688, 0.525, 0.712, 0.413, 0.606, 0.571, 0.26, 0.652, 0.358, 0.629, 0.66, 0.505, 0.518, 0.66, 0.789, 0.645, 0.66, 0.643, 0.646, 0.846, 0.521, 0.53, 0.522, 0.591, 0.505, 0.513, 0.496, 0.505, 0.51, 0.543, 0.647, 0.776, 0.575, 0.594, 0.754, 0.548, 0.668, 0.548, 0.749, 0.758, 0.871, 0.722, 0.643, 0.526, 0.695, 0.788, 0.697, 0.522, 0.833, 0.867, 0.688, 0.444, 0.509, 0.498, 0.687, 0.934, 0.824, 0.487, 0.777, 0.711, 0.528, 0.72, 0.635, 0.607, 0.552, 0.805, 0.701, 0.563, 0.557, 0.661, 0.754, 0.862, 0.791, 0.757, 0.51, 0.847, 0.565, 0.631, 0.464, 0.771, 0.511, 0.591, 0.515, 0.518, 0.852, 0.603, 0.61, 0.51, 0.6, 0.53, 0.476, 0.405, 0.575, 0.833, 0.817, 1.02, 0.977, 0.857, 0.626, 0.83, 0.913, 0.933, 0.411, 0.551, 0.783, 0.967, 0.705, 0.783, 0.552, 0.718, 0.587, 0.646, 0.826, 0.671, 0.788, 0.734, 0.709, 0.741, 0.469, 0.951, 0.891, 0.888, 0.678, 0.793, 1.018, 0.964, 0.817, 0.681, 0.884, 0.72, 0.427, 0.514, 0.602, 0.528, 0.874, 0.822, 0.845, 0.905, 0.78, 0.896, 0.865, 0.761, 0.908, 0.931, 0.663, 0.942, 0.888, 0.861, 0.928, 0.755, 0.786, 0.819, 0.797, 0.454, 0.565, 0.488, 0.321, 0.42, 0.508, 0.555, 0.286, 0.611, 0.633, 0.718, 0.462, 0.742, 0.858, 0.492, 0.757, 0.396, 0.83, 0.116, 0.772, 0.894, 0.531, 0.634, 0.844, 0.51, 0.409, 0.743, 0.467, 0.709, 0.576, 0.937, 0.474, 0.613, 0.878, 0.769, 0.857, 0.677, 0.87, 0.672, 0.69, 0.869, 0.602, 0.913, 0.924, 0.806, 0.758, 0.809, 0.717, 0.464, 0.772, 0.822, 0.881, 0.869, 0.904, 0.975, 0.852, 0.845, 0.807, 0.621, 0.844, 0.677, 0.679, 0.955, 0.941, 0.802, 0.93, 0.949, 0.658, 0.937, 0.706, 0.929, 0.748, 0.914, 0.135, 0.238, 0.718, 0.674, 0.443, 0.969, 1.018, 0.792, 0.989, 0.859, 0.13, 0.431, 0.339, 0.534, 0.612, 0.822, 0.468, 0.747, 0.393, 0.251, 0.649, 0.818, 0.627, 0.771, 0.873, 0.995, 0.829, 0.68, 0.903, 0.802, 0.614, 0.385, 0.371, 0.683, 0.642, 0.86, 0.98, 0.563, 0.771, 0.554, 0.597, 0.373, 0.678, 0.334, 0.372, 0.759, 0.352, 0.399, 0.781, 0.905, 0.334, 0.663, 0.796, 0.888, 0.487, 0.502, 0.419, 0.592, 0.42, 0.723, 0.164, 0.855, 0.541, 0.746, 0.728, 0.702, 0.086, 0.5, 0.892, 0.873, 0.162, 0.894, 0.723, 0.918, 0.62, 0.607]\n",
      "2428\n"
     ]
    }
   ],
   "source": [
    "#Parsing process + we are not thresholding the scores specific cutoff\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "Scores = []\n",
    "for seqs in SeqIO.parse(\"data.fasta\", \"fasta\"):\n",
    "    score = ((float(seqs.id[5:])))\n",
    "    Scores.append(score)\n",
    "    \n",
    "print(Scores)\n",
    "print(len(Scores))\n",
    "#the score represent the experimentally validated one that I put in the fasta header in order \n",
    "#to have an easy way to extract it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import GC\n",
    "from Bio import SeqIO\n",
    "seq_objects = SeqIO.parse(\"data.fasta\", \"fasta\")\n",
    "sequences = [seq for seq in seq_objects]\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = seq.seq\n",
    "    sequence = sequence.replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"u\", \"U\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#global GC content \n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import GC\n",
    "from Bio import SeqIO\n",
    "seq_objects = SeqIO.parse(\"data.fasta\", \"fasta\")\n",
    "sequences = [seq for seq in seq_objects]\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = seq.seq\n",
    "    gc_content = GC(sequence)\n",
    "    gc = round(gc_content,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 5 Extraction - Local GC Content (k=5)\n",
    "GC_I=[]\n",
    "GC_II=[]\n",
    "GC_III=[]\n",
    "GC_IV=[]\n",
    "GC_V=[]\n",
    "GC_VI=[]\n",
    "GC_VII=[]\n",
    "GC_VIII=[]\n",
    "GC_IX=[]\n",
    "GC_X=[]\n",
    "GC_XI=[]\n",
    "GC_XII=[]\n",
    "GC_XIII=[]\n",
    "GC_XIV=[]\n",
    "GC_XV=[]\n",
    "GC_XVI=[]\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sub1 = seq.seq[0:6]\n",
    "    sub2 = seq.seq[1:7]\n",
    "    sub3 = seq.seq[2:8]\n",
    "    sub3 = seq.seq[3:9]\n",
    "    sub4 = seq.seq[4:10]\n",
    "    sub5 = seq.seq[5:11]\n",
    "    sub6 = seq.seq[6:12]\n",
    "    sub7 = seq.seq[7:13]\n",
    "    sub8 = seq.seq[8:14]\n",
    "    sub9 = seq.seq[9:15]\n",
    "    sub10 = seq.seq[10:16]\n",
    "    sub11 = seq.seq[11:17]\n",
    "    sub12 = seq.seq[12:18]\n",
    "    sub13 = seq.seq[13:19]\n",
    "    sub14 = seq.seq[14:20]\n",
    "    sub15 = seq.seq[15:21]\n",
    "    sub16 = seq.seq[16:]\n",
    "    \n",
    "    \n",
    "    gc1 = round(GC(sub1),2)\n",
    "    gc2 = round(GC(sub2),2)\n",
    "    gc3 = round(GC(sub3),2)\n",
    "    gc4 = round(GC(sub4),2)\n",
    "    gc5 = round(GC(sub5),2)\n",
    "    gc6 = round(GC(sub6),2)\n",
    "    gc7 = round(GC(sub7),2)\n",
    "    gc8 = round(GC(sub8),2)\n",
    "    gc9 = round(GC(sub9),2)\n",
    "    gc10 = round(GC(sub10),2)\n",
    "    gc11 = round(GC(sub11),2)\n",
    "    gc12 = round(GC(sub12),2)\n",
    "    gc13 = round(GC(sub13),2)\n",
    "    gc14 = round(GC(sub14),2)\n",
    "    gc15 = round(GC(sub15),2)\n",
    "    gc16 = round(GC(sub16),2)\n",
    "    \n",
    "    GC_I.append(gc1)\n",
    "    GC_II.append(gc2)\n",
    "    GC_III.append(gc3)\n",
    "    GC_IV.append(gc4)\n",
    "    GC_V.append(gc5)\n",
    "    GC_VI.append(gc6)\n",
    "    GC_VII.append(gc7)\n",
    "    GC_VIII.append(gc8)\n",
    "    GC_IX.append(gc9)\n",
    "    GC_X.append(gc10)\n",
    "    GC_XI.append(gc11)\n",
    "    GC_XII.append(gc12)\n",
    "    GC_XIII.append(gc13)\n",
    "    GC_XIV.append(gc14)\n",
    "    GC_XV.append(gc15)\n",
    "    GC_XVI.append(gc16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 6 Extraction - siRNA Secondary Structure Î”G\n",
    "dgs=[]\n",
    "dgs3=[]\n",
    "dgs5=[]\n",
    "\n",
    "from seqfold import dg, Struct\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    five = sequence[0:7]\n",
    "    three = sequence[-7:]\n",
    "    delta_G = dg(sequence, temp = 25.0)\n",
    "   \n",
    "    dgs.append(delta_G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 7 Extraction - siRNA Melting Temperature\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "Tms=[]\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    Tm = mt.Tm_NN(sequence, check=True, strict=False, nn_table=mt.RNA_NN3)\n",
    "    Tms.append(round(Tm, 2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 9 Extraction - 5' A/U and G/C Richness (important in the antisense strand)\n",
    "five_prime_AU = []\n",
    "five_prime_GC = []\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    five_prime_AU_subseq = sequence[:7]\n",
    "    five_prime_GC_content = round(GC(five_prime_AU_subseq), 3)\n",
    "    five_prime_AU_content = round((100 - five_prime_GC_content), 3)\n",
    "    \n",
    "    five_prime_AU.append(five_prime_AU_content)\n",
    "    five_prime_GC.append(five_prime_GC_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 1 Extraction - Position-Specific Nucleotides + Class Label + Data Point IDs + Global GC Content\n",
    "seqs_ids=[]\n",
    "gc_contents=[]\n",
    "P1s=[]\n",
    "P2s=[]\n",
    "P3s=[]\n",
    "P4s=[]\n",
    "P5s=[]\n",
    "P6s=[]\n",
    "P7s=[]\n",
    "P8s=[]\n",
    "P9s=[]\n",
    "P10s=[]\n",
    "P11s=[]\n",
    "P12s=[]\n",
    "P13s=[]\n",
    "P14s=[]\n",
    "P15s=[]\n",
    "P16s=[]\n",
    "P17s=[]\n",
    "P18s=[]\n",
    "P19s=[]\n",
    "P20s=[]\n",
    "P21s=[]\n",
    "\n",
    "\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    gc_content = GC(sequence)\n",
    "    gc = round(gc_content,3)\n",
    "    P1 = sequence[0]\n",
    "    P2 = sequence[1]\n",
    "    P3 = sequence[2]\n",
    "    P4 = sequence[3]\n",
    "    P5 = sequence[4]\n",
    "    P6 = sequence[5]\n",
    "    P7 = sequence[6]\n",
    "    P8 = sequence[7]\n",
    "    P9 = sequence[8]\n",
    "    P10 = sequence[9]\n",
    "    P11 = sequence[10]\n",
    "    P12 = sequence[11]\n",
    "    P13 = sequence[12]\n",
    "    P14 = sequence[13]\n",
    "    P15 = sequence[14]\n",
    "    P16 = sequence[15]\n",
    "    P17 = sequence[16]\n",
    "    P18 = sequence[17]\n",
    "    P19 = sequence[18]\n",
    "    P20 = sequence[19]\n",
    "    P21 = sequence[20]\n",
    "    \n",
    "    \n",
    "    seqs_ids.append(seq_id)\n",
    "    gc_contents.append(gc)\n",
    "    P1s.append(P1)\n",
    "    P2s.append(P2)\n",
    "    P3s.append(P3)\n",
    "    P4s.append(P4)\n",
    "    P5s.append(P5)\n",
    "    P6s.append(P6)\n",
    "    P7s.append(P7)\n",
    "    P8s.append(P8)\n",
    "    P9s.append(P9)\n",
    "    P10s.append(P10)\n",
    "    P11s.append(P11)\n",
    "    P12s.append(P12)\n",
    "    P13s.append(P13)\n",
    "    P14s.append(P14)\n",
    "    P15s.append(P15)\n",
    "    P16s.append(P16)\n",
    "    P17s.append(P17)\n",
    "    P18s.append(P18)\n",
    "    P19s.append(P19)\n",
    "    P20s.append(P20)\n",
    "    P21s.append(P21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 8 Extraction - Dinucleotide frequency\n",
    "AC=[]\n",
    "AU=[]\n",
    "AG=[]\n",
    "AA=[]\n",
    "CA=[]\n",
    "CU=[]\n",
    "CG=[]\n",
    "CC=[]\n",
    "UU=[]\n",
    "UA=[]\n",
    "UG=[]\n",
    "UC=[]\n",
    "GU=[]\n",
    "GG=[]\n",
    "GA=[]\n",
    "GC=[]\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    ac=0\n",
    "    au=0\n",
    "    ag=0\n",
    "    aa=0\n",
    "    ca=0\n",
    "    cu=0\n",
    "    cg=0\n",
    "    cc=0\n",
    "    uu=0\n",
    "    ua=0\n",
    "    ug=0\n",
    "    uc=0\n",
    "    gu=0\n",
    "    gg=0\n",
    "    ga=0\n",
    "    gc=0\n",
    "    for i in range(len(sequence)-1):\n",
    "        dinuc = sequence[i:i+2]\n",
    "        if dinuc == \"AA\":\n",
    "            aa = aa + 1\n",
    "        elif dinuc == \"AC\":\n",
    "            ac = ac + 1\n",
    "        elif dinuc == \"AG\":\n",
    "            ag = ag +1\n",
    "        elif dinuc == \"AU\":\n",
    "            au = au +1\n",
    "        elif dinuc == \"CA\":\n",
    "            ca = ca + 1\n",
    "        elif dinuc == \"CC\":\n",
    "            cc = cc + 1\n",
    "        elif dinuc == \"CG\":\n",
    "            cg = cg +1\n",
    "        elif dinuc == \"CU\":\n",
    "            cu = cu +1\n",
    "        elif dinuc == \"GA\":\n",
    "            ga = ga + 1\n",
    "        elif dinuc == \"GC\":\n",
    "            gc = gc + 1\n",
    "        elif dinuc == \"GG\":\n",
    "            gg = gg +1\n",
    "        elif dinuc == \"GU\":\n",
    "            gu = gu +1\n",
    "        elif dinuc == \"UA\":\n",
    "            ua = ua + 1\n",
    "        elif dinuc == \"UC\":\n",
    "            uc = uc + 1\n",
    "        elif dinuc == \"UG\":\n",
    "            ug = ug +1\n",
    "        elif dinuc == \"UU\":\n",
    "            uu = uu +1\n",
    "    AA.append(round(((aa/20)*100), 3))\n",
    "    AC.append(round(((ac/20)*100), 3))\n",
    "    AG.append(round(((ag/20)*100), 3))\n",
    "    AU.append(round(((au/20)*100), 3))\n",
    "    CA.append(round(((ca/20)*100), 3))\n",
    "    CC.append(round(((cc/20)*100), 3))\n",
    "    CG.append(round(((cg/20)*100), 3))\n",
    "    CU.append(round(((cu/20)*100), 3))\n",
    "    GA.append(round(((aa/20)*100), 3))\n",
    "    GC.append(round(((gc/20)*100), 3))\n",
    "    GG.append(round(((gg/20)*100), 3))\n",
    "    GU.append(round(((gu/20)*100), 3))\n",
    "    UA.append(round(((ua/20)*100), 3))\n",
    "    UC.append(round(((uc/20)*100), 3))\n",
    "    UG.append(round(((ug/20)*100), 3))\n",
    "    UU.append(round(((uu/20)*100), 3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 8 Extraction - Trinucleotide frequency\n",
    "AAA=[]\n",
    "AAC=[]\n",
    "AAU=[]\n",
    "AAG=[]\n",
    "ACA=[]\n",
    "ACC=[]\n",
    "ACU=[]\n",
    "ACG=[]\n",
    "AUA=[]\n",
    "AUC=[]\n",
    "AUU=[]\n",
    "AUG=[]\n",
    "AGA=[]\n",
    "AGC=[]\n",
    "AGU=[]\n",
    "AGG=[]\n",
    "GGA=[]\n",
    "GGC=[]\n",
    "GGU=[]\n",
    "GGG=[]\n",
    "GCA=[]\n",
    "GCC=[]\n",
    "GCU=[]\n",
    "GCG=[]\n",
    "GUA=[]\n",
    "GUC=[]\n",
    "GUU=[]\n",
    "GAG=[]\n",
    "GAA=[]\n",
    "GAC=[]\n",
    "GAU=[]\n",
    "GUG=[]\n",
    "CAA=[]\n",
    "CAC=[]\n",
    "CAU=[]\n",
    "CAG=[]\n",
    "CCA=[]\n",
    "CCC=[]\n",
    "CCU=[]\n",
    "CCG=[]\n",
    "CGA=[]\n",
    "CGC=[]\n",
    "CGU=[]\n",
    "CGG=[]\n",
    "CUA=[]\n",
    "CUC=[]\n",
    "CUU=[]\n",
    "CUG=[]\n",
    "UAA=[]\n",
    "UAC=[]\n",
    "UAU=[]\n",
    "UAG=[]\n",
    "UCA=[]\n",
    "UCC=[]\n",
    "UCU=[]\n",
    "UCG=[]\n",
    "UGA=[]\n",
    "UGC=[]\n",
    "UGU=[]\n",
    "UGG=[]\n",
    "UUA=[]\n",
    "UUC=[]\n",
    "UUU=[]\n",
    "UUG=[]\n",
    "\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    aaa=0\n",
    "    aac=0\n",
    "    aau=0\n",
    "    aag=0\n",
    "    aca=0\n",
    "    acc=0\n",
    "    acu=0\n",
    "    acg=0\n",
    "    aua=0\n",
    "    auc=0\n",
    "    auu=0\n",
    "    aug=0\n",
    "    aga=0\n",
    "    agc=0\n",
    "    agu=0\n",
    "    agg=0\n",
    "    gga=0\n",
    "    ggc=0\n",
    "    ggu=0\n",
    "    ggg=0\n",
    "    gca=0\n",
    "    gcc=0\n",
    "    gcu=0\n",
    "    gcg=0\n",
    "    gua=0\n",
    "    guc=0\n",
    "    guu=0\n",
    "    gag=0\n",
    "    gaa=0\n",
    "    gac=0\n",
    "    gau=0\n",
    "    gug=0\n",
    "    caa=0\n",
    "    cac=0\n",
    "    cau=0\n",
    "    cag=0\n",
    "    cca=0\n",
    "    ccc=0\n",
    "    ccu=0\n",
    "    ccg=0\n",
    "    cga=0\n",
    "    cgc=0\n",
    "    cgu=0\n",
    "    cgg=0\n",
    "    cua=0\n",
    "    cuc=0\n",
    "    cuu=0\n",
    "    cug=0\n",
    "    uaa=0\n",
    "    uac=0\n",
    "    uau=0\n",
    "    uag=0\n",
    "    uca=0\n",
    "    ucc=0\n",
    "    ucu=0\n",
    "    ucg=0\n",
    "    uga=0\n",
    "    ugc=0\n",
    "    ugu=0\n",
    "    ugg=0\n",
    "    uua=0\n",
    "    uuc=0\n",
    "    uuu=0\n",
    "    uug=0\n",
    "    for i in range(len(sequence)-2):\n",
    "        dinuc = sequence[i:i+3]\n",
    "        if dinuc == \"AAA\":\n",
    "            aaa = aa + 1\n",
    "        elif dinuc == \"AAC\":\n",
    "            aac = aac + 1\n",
    "        elif dinuc == \"AAG\":\n",
    "            aag = aag +1\n",
    "        elif dinuc == \"AAU\":\n",
    "            aau = aau +1\n",
    "        elif dinuc == \"ACA\":\n",
    "            aca = aca + 1\n",
    "        elif dinuc == \"ACC\":\n",
    "            acc = acc + 1\n",
    "        elif dinuc == \"ACG\":\n",
    "            acg = acg +1\n",
    "        elif dinuc == \"ACU\":\n",
    "            acu = acu +1\n",
    "        elif dinuc == \"AGA\":\n",
    "            aga = aga + 1\n",
    "        elif dinuc == \"AGC\":\n",
    "            agc = agc + 1\n",
    "        elif dinuc == \"AGG\":\n",
    "            agg = agg +1\n",
    "        elif dinuc == \"AGU\":\n",
    "            agu = agu +1\n",
    "        elif dinuc == \"AUA\":\n",
    "            aua = aua + 1\n",
    "        elif dinuc == \"AUC\":\n",
    "            auc = auc + 1\n",
    "        elif dinuc == \"AUG\":\n",
    "            aug = aug +1\n",
    "        elif dinuc == \"AUU\":\n",
    "            auu = auu +1\n",
    "        elif dinuc == \"CCA\":\n",
    "            cca = cca + 1\n",
    "        elif dinuc == \"CCC\":\n",
    "            ccc = ccc + 1\n",
    "        elif dinuc == \"CCG\":\n",
    "            ccg = ccg +1\n",
    "        elif dinuc == \"CCU\":\n",
    "            ccu = ccu +1\n",
    "        elif dinuc == \"CGA\":\n",
    "            cga = cga + 1\n",
    "        elif dinuc == \"CGC\":\n",
    "            cgc = cgc + 1\n",
    "        elif dinuc == \"CGG\":\n",
    "            cgg = cgg +1\n",
    "        elif dinuc == \"CGU\":\n",
    "            cgu = cgu +1\n",
    "        elif dinuc == \"CUA\":\n",
    "            cua = cua + 1\n",
    "        elif dinuc == \"CUC\":\n",
    "            cuc = cuc + 1\n",
    "        elif dinuc == \"CUG\":\n",
    "            cug = cug +1\n",
    "        elif dinuc == \"CUU\":\n",
    "            cuu = cuu +1\n",
    "        elif dinuc == \"CAU\":\n",
    "            cau = cau +1\n",
    "        elif dinuc == \"CAA\":\n",
    "            caa = caa + 1\n",
    "        elif dinuc == \"CAC\":\n",
    "            cac = cac + 1\n",
    "        elif dinuc == \"CAG\":\n",
    "            cag = cag +1\n",
    "        elif dinuc == \"GCA\":\n",
    "            gca = gca + 1\n",
    "        elif dinuc == \"GCC\":\n",
    "            gcc = gcc + 1\n",
    "        elif dinuc == \"GCG\":\n",
    "            gcg = gcg +1\n",
    "        elif dinuc == \"GCU\":\n",
    "            gcu = gcu +1\n",
    "        elif dinuc == \"GGA\":\n",
    "            gga = gga + 1\n",
    "        elif dinuc == \"GGC\":\n",
    "            ggc = ggc + 1\n",
    "        elif dinuc == \"GGG\":\n",
    "            ggg = ggg +1\n",
    "        elif dinuc == \"GGU\":\n",
    "            ggu = ggu +1\n",
    "        elif dinuc == \"GUA\":\n",
    "            gua = gua + 1\n",
    "        elif dinuc == \"GUC\":\n",
    "            guc = guc + 1\n",
    "        elif dinuc == \"GUG\":\n",
    "            gug = gug +1\n",
    "        elif dinuc == \"GUU\":\n",
    "            guu = guu +1\n",
    "        elif dinuc == \"GAU\":\n",
    "            gau = gau +1\n",
    "        elif dinuc == \"GAA\":\n",
    "            gaa = gaa + 1\n",
    "        elif dinuc == \"GAC\":\n",
    "            gac = gac + 1\n",
    "        elif dinuc == \"GAG\":\n",
    "            gag = gag +1\n",
    "        elif dinuc == \"UCA\":\n",
    "            uca = uca + 1\n",
    "        elif dinuc == \"UCC\":\n",
    "            ucc = ucc + 1\n",
    "        elif dinuc == \"UCG\":\n",
    "            ucg = ucg +1\n",
    "        elif dinuc == \"UCU\":\n",
    "            ucu = ucu +1\n",
    "        elif dinuc == \"UGA\":\n",
    "            uga = uga + 1\n",
    "        elif dinuc == \"UGC\":\n",
    "            ugc = ugc + 1\n",
    "        elif dinuc == \"UGG\":\n",
    "            ugg = ugg +1\n",
    "        elif dinuc == \"UGU\":\n",
    "            ugu = ugu +1\n",
    "        elif dinuc == \"UUA\":\n",
    "            uua = uua + 1\n",
    "        elif dinuc == \"UUC\":\n",
    "            uuc = uuc + 1\n",
    "        elif dinuc == \"UUG\":\n",
    "            uug = uug +1\n",
    "        elif dinuc == \"UUU\":\n",
    "            uuu = uuu +1\n",
    "        elif dinuc == \"UAU\":\n",
    "            uau = uau +1\n",
    "        elif dinuc == \"UAA\":\n",
    "            uaa = uaa + 1\n",
    "        elif dinuc == \"UAC\":\n",
    "            uac = uac + 1\n",
    "        elif dinuc == \"UAG\":\n",
    "            uag = uag +1\n",
    "        \n",
    "    AAA.append(round(((aaa/19)*100), 3))\n",
    "    AAC.append(round(((aac/18)*100), 3))\n",
    "    AAG.append(round(((aag/19)*100), 3))\n",
    "    AAU.append(round(((aau/19)*100), 3))\n",
    "    ACA.append(round(((aca/19)*100), 3))\n",
    "    ACC.append(round(((acc/19)*100), 3))\n",
    "    ACG.append(round(((acg/19)*100), 3))\n",
    "    ACU.append(round(((acu/19)*100), 3))\n",
    "    AGA.append(round(((aga/19)*100), 3))\n",
    "    AGC.append(round(((agc/19)*100), 3))\n",
    "    AGG.append(round(((agg/19)*100), 3))\n",
    "    AGU.append(round(((agu/19)*100), 3))\n",
    "    AUA.append(round(((aua/19)*100), 3))\n",
    "    AUC.append(round(((auc/19)*100), 3))\n",
    "    AUG.append(round(((aug/19)*100), 3))\n",
    "    AUU.append(round(((auu/19)*100), 3))\n",
    "    \n",
    "    CAA.append(round(((caa/19)*100), 3))\n",
    "    CAC.append(round(((cac/19)*100), 3))\n",
    "    CAG.append(round(((cag/19)*100), 3))\n",
    "    CAU.append(round(((cau/19)*100), 3))\n",
    "    CCA.append(round(((cca/19)*100), 3))\n",
    "    CCC.append(round(((ccc/19)*100), 3))\n",
    "    CCG.append(round(((ccg/19)*100), 3))\n",
    "    CCU.append(round(((ccu/19)*100), 3))\n",
    "    CGA.append(round(((cga/19)*100), 3))\n",
    "    CGC.append(round(((cgc/19)*100), 3))\n",
    "    CGG.append(round(((cgg/19)*100), 3))\n",
    "    CGU.append(round(((cgu/19)*100), 3))\n",
    "    CUA.append(round(((cua/19)*100), 3))\n",
    "    CUC.append(round(((cuc/19)*100), 3))\n",
    "    CUG.append(round(((cug/19)*100), 3))\n",
    "    CUU.append(round(((cuu/19)*100), 3))\n",
    "    \n",
    "    GAA.append(round(((gaa/19)*100),3))\n",
    "    GAC.append(round(((gac/19)*100),3))\n",
    "    GAG.append(round(((gag/19)*100),3))\n",
    "    GAU.append(round(((gau/19)*100),3))\n",
    "    GCA.append(round(((gca/19)*100),3))\n",
    "    GCC.append(round(((gcc/19)*100),3))\n",
    "    GCG.append(round(((gcg/19)*100),3))\n",
    "    GCU.append(round(((gcu/19)*100),3))\n",
    "    GGA.append(round(((gga/19)*100),3))\n",
    "    GGC.append(round(((ggc/19)*100),3))\n",
    "    GGG.append(round(((ggg/19)*100),3))\n",
    "    GGU.append(round(((ggu/19)*100),3))\n",
    "    GUA.append(round(((gua/19)*100),3))\n",
    "    GUC.append(round(((guc/19)*100),3))\n",
    "    GUG.append(round(((gug/19)*100),3))\n",
    "    GUU.append(round(((guu/19)*100),3))\n",
    "    \n",
    "    UAA.append(round(((uaa/19)*100),3))\n",
    "    UAC.append(round(((uac/19)*100),3))\n",
    "    UAG.append(round(((uag/19)*100),3))\n",
    "    UAU.append(round(((uau/19)*100),3))\n",
    "    UCA.append(round(((uca/19)*100),3))\n",
    "    UCC.append(round(((ucc/19)*100),3))\n",
    "    UCG.append(round(((ucg/19)*100),3))\n",
    "    UCU.append(round(((ucu/19)*100),3))\n",
    "    UGA.append(round(((uga/19)*100),3))\n",
    "    UGC.append(round(((ugc/19)*100),3))\n",
    "    UGG.append(round(((ugg/19)*100),3))\n",
    "    UGU.append(round(((ugu/19)*100),3))\n",
    "    UUA.append(round(((uua/19)*100),3))\n",
    "    UUC.append(round(((uuc/19)*100),3))\n",
    "    UUG.append(round(((uug/19)*100),3))\n",
    "    UUU.append(round(((uuu/19)*100),3))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Features Group 9 Extraction - 5' and 3' end (7 nt)\n",
    "five_prime_7nt = []\n",
    "three_prime_7nt = []\n",
    "\n",
    "for seq in sequences:\n",
    "    seq_id = seq.id\n",
    "    sequence = (seq.seq).replace(\"t\", \"U\")\n",
    "    sequence = sequence.replace(\"a\", \"A\")\n",
    "    sequence = sequence.replace(\"g\", \"G\")\n",
    "    sequence = sequence.replace(\"c\", \"C\")\n",
    "    five = sequence[:7]\n",
    "    three = sequence[-7:]\n",
    "    five_prime_7nt.append(five)\n",
    "    three_prime_7nt.append(three)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:109: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['CUU%']=CUU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:110: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['CUG%']=CUG\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:111: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UAA%']=UAA\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:112: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UAC%']=UAC\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:113: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UAU%']=UAU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:114: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UAG%']=UAG\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:115: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UCA%']=UCA\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:116: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UCC%']=UCC\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:117: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UCU%']=UCU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:118: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UCG%']=UCG\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:119: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UGA%']=UGA\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:120: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UGC%']=UGC\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:121: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UGU%']=UGU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:122: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UGG%']=UGG\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:123: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UUA%']=UUA\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:124: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UUC%']=UUC\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:125: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UUU%']=UUU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:126: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['UUG%']=UUG\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:128: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['five_prime_GC']=five_prime_GC\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:129: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['five_prime_AU']=five_prime_AU\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:131: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['Global_Tm']=Tms  #using termodynamics neighbor\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:132: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['Global_Î”G']=dgs  #at 25Â°C\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:134: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['five_prime_seq']=five_prime_7nt\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:135: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['three_prime_seq']=three_prime_7nt\n",
      "/var/folders/_k/k2b15r4d7_qd3_2r3f5bmq300000gn/T/ipykernel_93327/3429556490.py:137: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  dataframe['Label']=Scores\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "dataframe=pd.DataFrame()\n",
    "#sequences id\n",
    "#dataframe['Sequence_ID']=seqs_ids\n",
    "#position specific nucleotides\n",
    "dataframe['P1']=P1s\n",
    "dataframe['P2']=P2s\n",
    "dataframe['P3']=P3s\n",
    "dataframe['P4']=P4s\n",
    "dataframe['P5']=P5s\n",
    "dataframe['P6']=P6s\n",
    "dataframe['P7']=P7s\n",
    "dataframe['P8']=P8s\n",
    "dataframe['P9']=P9s\n",
    "dataframe['P10']=P10s\n",
    "dataframe['P11']=P11s\n",
    "dataframe['P12']=P12s\n",
    "dataframe['P13']=P13s\n",
    "dataframe['P14']=P14s\n",
    "dataframe['P15']=P15s\n",
    "dataframe['P16']=P16s\n",
    "dataframe['P17']=P17s\n",
    "dataframe['P18']=P18s\n",
    "dataframe['P19']=P19s\n",
    "dataframe['P20']=P20s\n",
    "dataframe['P21']=P21s\n",
    "#global and local gc content, with k=5 for local\n",
    "dataframe['GC_Content']=gc_contents\n",
    "dataframe['GC_1']=GC_I\n",
    "dataframe['GC_2']=GC_II\n",
    "dataframe['GC_3']=GC_III\n",
    "dataframe['GC_4']=GC_IV\n",
    "dataframe['GC_5']=GC_V\n",
    "dataframe['GC_6']=GC_VI\n",
    "dataframe['GC_7']=GC_VII\n",
    "dataframe['GC_8']=GC_VIII\n",
    "dataframe['GC_9']=GC_IX\n",
    "dataframe['GC_10']=GC_X\n",
    "dataframe['GC_11']=GC_XI\n",
    "dataframe['GC_12']=GC_XII\n",
    "dataframe['GC_13']=GC_XIII\n",
    "dataframe['GC_14']=GC_XIV\n",
    "dataframe['GC_15']=GC_XV\n",
    "dataframe['GC_16']=GC_XVI\n",
    "#dinucleotides content\n",
    "dataframe['AA%']=AA\n",
    "dataframe['AC%']=AC\n",
    "dataframe['AG%']=AG\n",
    "dataframe['AU%']=AU\n",
    "dataframe['CA%']=CA\n",
    "dataframe['CC%']=CC\n",
    "dataframe['CG%']=CG\n",
    "dataframe['CU%']=CU\n",
    "dataframe['GG%']=GG\n",
    "dataframe['GA%']=GA\n",
    "dataframe['GC%']=GC\n",
    "dataframe['GU%']=GU\n",
    "dataframe['UA%']=UA\n",
    "dataframe['UC%']=UC\n",
    "dataframe['UG%']=UG\n",
    "dataframe['UU%']=UU\n",
    "#trinucleotides content\n",
    "dataframe['AAA%']=AAA\n",
    "dataframe['AAC%']=AAC\n",
    "dataframe['AAU%']=AAU\n",
    "dataframe['AAG%']=AAG\n",
    "dataframe['ACA%']=ACA\n",
    "dataframe['ACC%']=ACC\n",
    "dataframe['ACU%']=ACU\n",
    "dataframe['ACG%']=ACG\n",
    "dataframe['AUA%']=AUA\n",
    "dataframe['AUC%']=AUC\n",
    "dataframe['AUU%']=AUU\n",
    "dataframe['AUG%']=AUG\n",
    "dataframe['AGA%']=AGA\n",
    "dataframe['AGC%']=AGC\n",
    "dataframe['AGU%']=AGU\n",
    "dataframe['AGG%']=AGG\n",
    "dataframe['GGA%']=GGA\n",
    "dataframe['GGC%']=GGC\n",
    "dataframe['GGU%']=GGU\n",
    "dataframe['GGG%']=GGG\n",
    "dataframe['GCA%']=GCA\n",
    "dataframe['GCC%']=GCC\n",
    "dataframe['GCU%']=GCU\n",
    "dataframe['GCG%']=GCG\n",
    "dataframe['GUA%']=GUA\n",
    "dataframe['GUC%']=GUC\n",
    "dataframe['GUU%']=GUU\n",
    "dataframe['GAG%']=GAG\n",
    "dataframe['GAA%']=GAA\n",
    "dataframe['GAC%']=GAC\n",
    "dataframe['GAU%']=GAU\n",
    "dataframe['GUG%']=GUG\n",
    "dataframe['CAA%']=CAA\n",
    "dataframe['CAC%']=CAC\n",
    "dataframe['CAU%']=CAU\n",
    "dataframe['CAG%']=CAG\n",
    "dataframe['CCA%']=CCA\n",
    "dataframe['CCC%']=CCC\n",
    "dataframe['CCU%']=CCU\n",
    "dataframe['CCG%']=CCG\n",
    "dataframe['CGA%']=CGA\n",
    "dataframe['CGC%']=CGC\n",
    "dataframe['CGU%']=CGU\n",
    "dataframe['CGG%']=CGG\n",
    "dataframe['CUA%']=CUA\n",
    "dataframe['CUC%']=CUC\n",
    "dataframe['CUU%']=CUU\n",
    "dataframe['CUG%']=CUG\n",
    "dataframe['UAA%']=UAA\n",
    "dataframe['UAC%']=UAC\n",
    "dataframe['UAU%']=UAU\n",
    "dataframe['UAG%']=UAG\n",
    "dataframe['UCA%']=UCA\n",
    "dataframe['UCC%']=UCC\n",
    "dataframe['UCU%']=UCU\n",
    "dataframe['UCG%']=UCG\n",
    "dataframe['UGA%']=UGA\n",
    "dataframe['UGC%']=UGC\n",
    "dataframe['UGU%']=UGU\n",
    "dataframe['UGG%']=UGG\n",
    "dataframe['UUA%']=UUA\n",
    "dataframe['UUC%']=UUC\n",
    "dataframe['UUU%']=UUU\n",
    "dataframe['UUG%']=UUG\n",
    "#5' GC, AU content\n",
    "dataframe['five_prime_GC']=five_prime_GC\n",
    "dataframe['five_prime_AU']=five_prime_AU\n",
    "#thermodynamics profile\n",
    "dataframe['Global_Tm']=Tms  #using termodynamics neighbor\n",
    "dataframe['Global_Î”G']=dgs  #at 25Â°C\n",
    "#5' and 3' end sequences with k=7\n",
    "dataframe['five_prime_seq']=five_prime_7nt\n",
    "dataframe['three_prime_seq']=three_prime_7nt\n",
    "#classes' label\n",
    "dataframe['Label']=Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     P1 P2 P3 P4 P5 P6 P7 P8 P9 P10  ...    UUC%    UUU%   UUG% five_prime_GC  \\\n",
      "0     C  U  A  A  U  A  U  G  U   U  ...   0.000   5.263  5.263        14.286   \n",
      "1     A  A  U  A  U  G  U  U  A   A  ...   0.000   5.263  5.263        14.286   \n",
      "2     G  A  U  U  U  A  U  A  C   A  ...  10.526  10.526  0.000        14.286   \n",
      "3     C  A  A  U  U  C  C  U  U   U  ...  10.526  15.789  0.000        42.857   \n",
      "4     C  A  G  A  C  C  A  A  A   A  ...   0.000   0.000  0.000        57.143   \n",
      "...  .. .. .. .. .. .. .. .. ..  ..  ...     ...     ...    ...           ...   \n",
      "2423  A  A  A  A  C  C  C  A  C   C  ...   0.000   0.000  0.000        42.857   \n",
      "2424  G  A  U  G  A  U  C  A  G   A  ...   0.000   0.000  0.000        42.857   \n",
      "2425  A  U  C  A  G  A  A  G  C   U  ...   0.000   0.000  0.000        28.571   \n",
      "2426  G  A  A  G  C  U  G  A  A   C  ...   0.000   0.000  5.263        57.143   \n",
      "2427  G  A  U  G  A  U  C  C  C   G  ...   0.000   0.000  0.000        42.857   \n",
      "\n",
      "     five_prime_AU Global_Tm Global_Î”G         five_prime_seq  \\\n",
      "0           85.714     46.45       1.5  (C, U, A, A, U, A, U)   \n",
      "1           85.714     46.80       0.6  (A, A, U, A, U, G, U)   \n",
      "2           85.714     53.21      -1.6  (G, A, U, U, U, A, U)   \n",
      "3           57.143     52.19       0.7  (C, A, A, U, U, C, C)   \n",
      "4           42.857     52.69       2.9  (C, A, G, A, C, C, A)   \n",
      "...            ...       ...       ...                    ...   \n",
      "2423        57.143     66.66      -1.5  (A, A, A, A, C, C, C)   \n",
      "2424        57.143     64.48      -2.4  (G, A, U, G, A, U, C)   \n",
      "2425        71.429     67.66      -3.3  (A, U, C, A, G, A, A)   \n",
      "2426        42.857     67.90       0.3  (G, A, A, G, C, U, G)   \n",
      "2427        57.143     73.10      -2.2  (G, A, U, G, A, U, C)   \n",
      "\n",
      "            three_prime_seq  Label  \n",
      "0     (G, A, U, U, U, A, U)  0.462  \n",
      "1     (U, U, U, A, U, A, C)  0.384  \n",
      "2     (C, U, U, U, C, A, A)  0.514  \n",
      "3     (U, U, U, A, U, C, U)  0.364  \n",
      "4     (A, U, A, A, G, A, A)  0.522  \n",
      "...                     ...    ...  \n",
      "2423  (A, A, A, G, U, A, G)  0.894  \n",
      "2424  (G, A, A, C, U, A, C)  0.723  \n",
      "2425  (U, A, C, U, C, G, G)  0.918  \n",
      "2426  (C, G, G, U, U, G, A)  0.620  \n",
      "2427  (U, C, C, A, G, C, U)  0.607  \n",
      "\n",
      "[2428 rows x 125 columns]\n"
     ]
    }
   ],
   "source": [
    "print(dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputfile='/Users/christian/Desktop/siRNA ML/siRNA_features.csv'\n",
    "dataframe.to_csv(outputfile,index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Notebook Created By: Christian Mandelli, Oregon State University*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
